source/common/upstream/load_balancer_impl.h

```diff
@@ -415,6 +428,30 @@ class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase {
   // Scheduler for each valid HostsSource.
   absl::node_hash_map<HostsSource, Scheduler, HostsSourceHash> scheduler_;
   Common::CallbackHandlePtr priority_update_cb_;
+  Common::CallbackHandlePtr member_update_cb_;
+
+protected:
+  // Slow start related configs
+  const std::chrono::milliseconds slow_start_window_;
+  double time_bias_{1.0};
+  double aggression_{1.0};
+  const std::unique_ptr<Runtime::Double> time_bias_runtime_;
+  const std::unique_ptr<Runtime::Double> aggression_runtime_;
+  TimeSource& time_source_;
+  struct OrderByCreateDateDesc {
+    bool operator()(const HostSharedPtr l, const HostSharedPtr r) const {
+      return (l->creationTime() < r->creationTime()) ||
+             (l->address()->asString() < r->address()->asString());
+    }
+  };
+  // Used exclusively for:
+  //    - checking if at least one host is in slow start;
+  //    - ordering of hosts by creation date in ascending order.
+  // Not meant to check if given host is in slow start as check relies on current time.
+  // That check could be done inline by comparing the diff of current time and host creation date
+  // against slow start window and by checking if hosts adheres to endpoint warming policy.
+  std::shared_ptr<absl::btree_set<HostSharedPtr, OrderByCreateDateDesc>> hosts_in_slow_start_;
```

Comment:
+1 to a timer and bool state. I think that's pretty elegant.

Reply:
@wbpcode Just to make sure I understand the suggestion. Keeping only one timer of latest added host is not enough to reliably track if any host is in slow start . Is the suggestion to replace btree with a collection of timers and one bool variable, in that case what is the need for bool variable? Or did you mean storing a map of <timer, bool_state>, in that case the presence of timer in the map can be an indication if some host is in slow start, no need to store the bool state as well. (once timer expires, it gets removed from collection)

Reply:
@nezdolik 🤔 I think it is enough to just create a timer for the last host added. When this timer expires, all hosts must be outside the slow start time window.
Can you explain why you think one timer is not enough?

Reply:
@wbpcode in my first implementation of slow start i was using timestamp of latest added host, but then realised that it will not cover some use cases, for example:
![Image from iOS (7)](https://user-images.githubusercontent.com/14430883/126291494-3c67e1b1-374c-41ee-9c67-a90c09e71944.jpg)

Reply:
https://github.com/envoyproxy/envoy/pull/13176#issuecomment-704148676

Reply:
@nezdolik I understand. But I think that in this case you said, the price you pay is only in a slow_start_window time, the cluster is mistaken for being in a slow start state, and some additional calculations are required. This is not unacceptable overhead. And the implementation is cleaner and simpler.
To be honest, envoy is complex enough, and it’s not a bad thing to be simple and effective.

Reply:
i now need to reread impl code again after a 2 month pause. The price you pay would be that in unlucky case Envoy would consider there are no hosts in slow start even if there are some (like host A in example above), edf creation would be skipped and the weights will not be recalculated. wdyt @wbpcode